File: mobile/ios/VPN/ALADDINVPNClient.swift
//
//  ALADDINVPNClient.swift
//  ALADDIN Security
//
//  Created by AI Assistant on 2025-01-27.
//  Copyright © 2025 ALADDIN Security. All rights reserved.
//

import Foundation
import NetworkExtension
import UIKit

// MARK: - VPN Status Enum
enum VPNStatus: String, CaseIterable {
    case disconnected = "disconnected"
    case connecting = "connecting"
    case connected = "connected"
    case disconnecting = "disconnecting"
    case error = "error"
    
    var displayName: String {
        switch self {
        case .disconnected: return "Отключен"
        case .connecting: return "Подключение..."
        case .connected: return "Подключен"
        case .disconnecting: return "Отключение..."
        case .error: return "Ошибка"
        }
    }
    
    var color: UIColor {
        switch self {
        case .disconnected: return UIColor.systemGray
        case .connecting: return UIColor.systemYellow
        case .connected: return UIColor.systemGreen
        case .disconnecting: return UIColor.systemOrange
        case .error: return UIColor.systemRed
        }
    }
}

// MARK: - VPN Protocol Enum
enum VPNProtocol: String, CaseIterable {
    case wireguard = "wireguard"
    case openvpn = "openvpn"
    case shadowsocks = "shadowsocks"
    case v2ray = "v2ray"
    
    var displayName: String {
        switch self {
        case .wireguard: return "WireGuard"
        case .openvpn: return "OpenVPN"
        case .shadowsocks: return "Shadowsocks"
        case .v2ray: return "V2Ray"
        }
    }
}

// MARK: - VPN Server Model
struct VPNServer: Codable, Identifiable {
    let id: String
    let name: String
    let country: String
    let flag: String
    let ip: String
    let port: Int
    let protocol: VPNProtocol
    let isAvailable: Bool
    let performanceScore: Double
    let ping: Int
    let load: Int
    
    var displayName: String {
        return "\(flag) \(name)"
    }
    
    var statusText: String {
        return "\(ping)ms • \(load)% нагрузка"
    }
}

// MARK: - Connection Info
struct ConnectionInfo {
    let server: VPNServer
    let startTime: Date
    var bytesSent: Int64
    var bytesReceived: Int64
    let status: VPNStatus
    
    var connectionTime: TimeInterval {
        return Date().timeIntervalSince(startTime)
    }
    
    var speed: (download: Double, upload: Double) {
        let time = connectionTime
        guard time > 0 else { return (0, 0) }
        
        let downloadSpeed = Double(bytesReceived) / time / 1024 // KB/s
        let uploadSpeed = Double(bytesSent) / time / 1024 // KB/s
        
        return (downloadSpeed, uploadSpeed)
    }
}

// MARK: - ALADDIN VPN Client
class ALADDINVPNClient: ObservableObject {
    
    // MARK: - Published Properties
    @Published var status: VPNStatus = .disconnected
    @Published var currentConnection: ConnectionInfo?
    @Published var availableServers: [VPNServer] = []
    @Published var isConnecting = false
    @Published var errorMessage: String?
    
    // MARK: - Private Properties
    private var vpnManager: NEVPNManager?
    private var connectionHistory: [ConnectionInfo] = []
    private let apiClient: ALADDINAPIClient
    
    // MARK: - Initialization
    init(apiClient: ALADDINAPIClient) {
        self.apiClient = apiClient
        setupVPNManager()
        loadServers()
    }
    
    // MARK: - Setup
    private func setupVPNManager() {
        NEVPNManager.shared().loadFromPreferences { [weak self] error in
            if let error = error {
                print("Ошибка загрузки VPN настроек: \(error)")
                return
            }
            
            DispatchQueue.main.async {
                self?.vpnManager = NEVPNManager.shared()
                self?.updateStatus()
            }
        }
    }
    
    private func loadServers() {
        // Загружаем серверы из API
        Task {
            do {
                let servers = try await apiClient.getVPNServers()
                await MainActor.run {
                    self.availableServers = servers
                }
            } catch {
                print("Ошибка загрузки серверов: \(error)")
                // Используем серверы по умолчанию
                await MainActor.run {
                    self.loadDefaultServers()
                }
            }
        }
    }
    
    private func loadDefaultServers() {
        availableServers = [
            VPNServer(
                id: "sg-1",
                name: "Сингапур",
                country: "SG",
                flag: "🇸🇬",
                ip: "192.168.2.10",
                port: 443,
                protocol: .shadowsocks,
                isAvailable: true,
                performanceScore: 95.0,
                ping: 25,
                load: 15
            ),
            VPNServer(
                id: "de-1",
                name: "Германия",
                country: "DE",
                flag: "🇩🇪",
                ip: "192.168.2.11",
                port: 443,
                protocol: .v2ray,
                isAvailable: true,
                performanceScore: 92.0,
                ping: 45,
                load: 25
            ),
            VPNServer(
                id: "hk-1",
                name: "Гонконг",
                country: "HK",
                flag: "🇭🇰",
                ip: "192.168.2.12",
                port: 443,
                protocol: .shadowsocks,
                isAvailable: true,
                performanceScore: 88.0,
                ping: 35,
                load: 20
            ),
            VPNServer(
                id: "jp-1",
                name: "Япония",
                country: "JP",
                flag: "🇯🇵",
                ip: "192.168.2.13",
                port: 51820,
                protocol: .wireguard,
                isAvailable: true,
                performanceScore: 90.0,
                ping: 40,
                load: 30
            ),
            VPNServer(
                id: "us-1",
                name: "США",
                country: "US",
                flag: "🇺🇸",
                ip: "192.168.2.14",
                port: 1194,
                protocol: .openvpn,
                isAvailable: true,
                performanceScore: 85.0,
                ping: 80,
                load: 40
            ),
            VPNServer(
                id: "ca-1",
                name: "Канада",
                country: "CA",
                flag: "🇨🇦",
                ip: "192.168.2.15",
                port: 51820,
                protocol: .wireguard,
                isAvailable: true,
                performanceScore: 87.0,
                ping: 75,
                load: 35
            )
        ]
    }
    
    // MARK: - Public Methods
    
    /// Подключение к VPN серверу
    func connect(to server: VPNServer? = nil) async {
        guard !isConnecting else { return }
        
        let targetServer = server ?? selectBestServer()
        guard let targetServer = targetServer else {
            await MainActor.run {
                errorMessage = "Нет доступных серверов"
            }
            return
        }
        
        await MainActor.run {
            isConnecting = true
            status = .connecting
            errorMessage = nil
        }
        
        do {
            try await performConnection(to: targetServer)
            
            await MainActor.run {
                currentConnection = ConnectionInfo(
                    server: targetServer,
                    startTime: Date(),
                    bytesSent: 0,
                    bytesReceived: 0,
                    status: .connected
                )
                status = .connected
                isConnecting = false
                connectionHistory.append(currentConnection!)
            }
            
        } catch {
            await MainActor.run {
                status = .error
                isConnecting = false
                errorMessage = error.localizedDescription
            }
        }
    }
    
    /// Отключение от VPN
    func disconnect() async {
        guard status != .disconnected else { return }
        
        await MainActor.run {
            status = .disconnecting
        }
        
        do {
            try await performDisconnection()
            
            await MainActor.run {
                currentConnection = nil
                status = .disconnected
            }
            
        } catch {
            await MainActor.run {
                status = .error
                errorMessage = error.localizedDescription
            }
        }
    }
    
    /// Выбор лучшего сервера
    func selectBestServer() -> VPNServer? {
        let availableServers = availableServers.filter { $0.isAvailable }
        guard !availableServers.isEmpty else { return nil }
        
        // Выбираем сервер с лучшей производительностью
        return availableServers.max { $0.performanceScore < $1.performanceScore }
    }
    
    /// Получение серверов для быстрого подключения
    func getQuickConnectServers() -> [VPNServer] {
        return Array(availableServers
            .filter { $0.isAvailable }
            .sorted { $0.ping < $1.ping }
            .prefix(4))
    }
    
    /// Обновление статистики трафика
    func updateTraffic(bytesSent: Int64, bytesReceived: Int64) {
        guard var connection = currentConnection else { return }
        
        connection.bytesSent += bytesSent
        connection.bytesReceived += bytesReceived
        
        currentConnection = connection
    }
    
    /// Получение сводки подключения
    func getConnectionSummary() -> [String: Any] {
        guard let connection = currentConnection else {
            return [
                "isConnected": false,
                "statusText": status.displayName,
                "serverInfo": [String: Any](),
                "connectionTime": 0.0,
                "speed": ["download": 0.0, "upload": 0.0],
                "dataUsage": ["sent": 0, "received": 0, "total": 0]
            ]
        }
        
        let speed = connection.speed
        
        return [
            "isConnected": status == .connected,
            "statusText": status.displayName,
            "serverInfo": [
                "id": connection.server.id,
                "name": connection.server.name,
                "country": connection.server.country,
                "flag": connection.server.flag
            ],
            "connectionTime": connection.connectionTime,
            "speed": [
                "download": speed.download,
                "upload": speed.upload
            ],
            "dataUsage": [
                "sent": connection.bytesSent,
                "received": connection.bytesReceived,
                "total": connection.bytesSent + connection.bytesReceived
            ]
        ]
    }
    
    // MARK: - Private Methods
    
    private func performConnection(to server: VPNServer) async throws {
        // Здесь будет реальная логика подключения к VPN
        // Пока что симулируем подключение
        
        try await Task.sleep(nanoseconds: 2_000_000_000) // 2 секунды
        
        // Симулируем успешное подключение
        // В реальной реализации здесь будет настройка NEVPNManager
    }
    
    private func performDisconnection() async throws {
        // Здесь будет реальная логика отключения от VPN
        // Пока что симулируем отключение
        
        try await Task.sleep(nanoseconds: 1_000_000_000) // 1 секунда
        
        // В реальной реализации здесь будет отключение NEVPNManager
    }
    
    private func updateStatus() {
        guard let manager = vpnManager else { return }
        
        switch manager.connection.status {
        case .connected:
            status = .connected
        case .connecting:
            status = .connecting
        case .disconnecting:
            status = .disconnecting
        case .disconnected:
            status = .disconnected
        case .invalid:
            status = .error
        @unknown default:
            status = .error
        }
    }
}

// MARK: - ALADDIN API Client Protocol
protocol ALADDINAPIClient {
    func getVPNServers() async throws -> [VPNServer]
}

File: mobile/android/VPN/ALADDINVPNClient.kt
package com.aladdin.security.vpn

import android.content.Context
import android.net.VpnService
import android.os.ParcelFileDescriptor
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import java.util.concurrent.atomic.AtomicBoolean

/**
 * ALADDIN VPN Client для Android
 * Обеспечивает безопасное подключение к VPN серверам
 */
class ALADDINVPNClient(private val context: Context) {
    
    // MARK: - Enums
    enum class VPNStatus(val displayName: String) {
        DISCONNECTED("Отключен"),
        CONNECTING("Подключение..."),
        CONNECTED("Подключен"),
        DISCONNECTING("Отключение..."),
        ERROR("Ошибка")
    }
    
    enum class VPNProtocol(val displayName: String) {
        WIREGUARD("WireGuard"),
        OPENVPN("OpenVPN"),
        SHADOWSOCKS("Shadowsocks"),
        V2RAY("V2Ray")
    }
    
    // MARK: - Data Classes
    data class VPNServer(
        val id: String,
        val name: String,
        val country: String,
        val flag: String,
        val ip: String,
        val port: Int,
        val protocol: VPNProtocol,
        val isAvailable: Boolean,
        val performanceScore: Double,
        val ping: Int,
        val load: Int
    ) {
        val displayName: String get() = "$flag $name"
        val statusText: String get() = "${ping}ms • ${load}% нагрузка"
    }
    
    data class ConnectionInfo(
        val server: VPNServer,
        val startTime: Long,
        var bytesSent: Long,
        var bytesReceived: Long,
        val status: VPNStatus
    ) {
        val connectionTime: Long get() = System.currentTimeMillis() - startTime
        
        val speed: Pair<Double, Double> get() {
            val time = connectionTime / 1000.0
            if (time <= 0) return Pair(0.0, 0.0)
            
            val downloadSpeed = bytesReceived / time / 1024 // KB/s
            val uploadSpeed = bytesSent / time / 1024 // KB/s
            
            return Pair(downloadSpeed, uploadSpeed)
        }
    }
    
    // MARK: - StateFlow Properties
    private val _status = MutableStateFlow(VPNStatus.DISCONNECTED)
    val status: StateFlow<VPNStatus> = _status.asStateFlow()
    
    private val _currentConnection = MutableStateFlow<ConnectionInfo?>(null)
    val currentConnection: StateFlow<ConnectionInfo?> = _currentConnection.asStateFlow()
    
    private val _availableServers = MutableStateFlow<List<VPNServer>>(emptyList())
    val availableServers: StateFlow<List<VPNServer>> = _availableServers.asStateFlow()
    
    private val _isConnecting = MutableStateFlow(false)
    val isConnecting: StateFlow<Boolean> = _isConnecting.asStateFlow()
    
    private val _errorMessage = MutableStateFlow<String?>(null)
    val errorMessage: StateFlow<String?> = _errorMessage.asStateFlow()
    
    // MARK: - Private Properties
    private val scope = CoroutineScope(Dispatchers.Main + SupervisorJob())
    private val connectionHistory = mutableListOf<ConnectionInfo>()
    private val isConnected = AtomicBoolean(false)
    private var vpnService: VpnService? = null
    private var vpnInterface: ParcelFileDescriptor? = null
    
    // MARK: - Initialization
    init {
        loadServers()
    }
    
    // MARK: - Public Methods
    
    /**
     * Подключение к VPN серверу
     */
    suspend fun connect(server: VPNServer? = null) {
        if (_isConnecting.value) return
        
        val targetServer = server ?: selectBestServer()
        if (targetServer == null) {
            _errorMessage.value = "Нет доступных серверов"
            return
        }
        
        _isConnecting.value = true
        _status.value = VPNStatus.CONNECTING
        _errorMessage.value = null
        
        try {
            performConnection(targetServer)
            
            val connection = ConnectionInfo(
                server = targetServer,
                startTime = System.currentTimeMillis(),
                bytesSent = 0,
                bytesReceived = 0,
                status = VPNStatus.CONNECTED
            )
            
            _currentConnection.value = connection
            _status.value = VPNStatus.CONNECTED
            _isConnecting.value = false
            connectionHistory.add(connection)
            isConnected.set(true)
            
        } catch (e: Exception) {
            _status.value = VPNStatus.ERROR
            _isConnecting.value = false
            _errorMessage.value = e.message
        }
    }
    
    /**
     * Отключение от VPN
     */
    suspend fun disconnect() {
        if (_status.value == VPNStatus.DISCONNECTED) return
        
        _status.value = VPNStatus.DISCONNECTING
        
        try {
            performDisconnection()
            
            _currentConnection.value = null
            _status.value = VPNStatus.DISCONNECTED
            isConnected.set(false)
            
        } catch (e: Exception) {
            _status.value = VPNStatus.ERROR
            _errorMessage.value = e.message
        }
    }
    
    /**
     * Выбор лучшего сервера
     */
    fun selectBestServer(): VPNServer? {
        val availableServers = _availableServers.value.filter { it.isAvailable }
        if (availableServers.isEmpty()) return null
        
        // Выбираем сервер с лучшей производительностью
        return availableServers.maxByOrNull { it.performanceScore }
    }
    
    /**
     * Получение серверов для быстрого подключения
     */
    fun getQuickConnectServers(): List<VPNServer> {
        return _availableServers.value
            .filter { it.isAvailable }
            .sortedBy { it.ping }
            .take(4)
    }
    
    /**
     * Обновление статистики трафика
     */
    fun updateTraffic(bytesSent: Long, bytesReceived: Long) {
        val connection = _currentConnection.value ?: return
        
        val updatedConnection = connection.copy(
            bytesSent = connection.bytesSent + bytesSent,
            bytesReceived = connection.bytesReceived + bytesReceived
        )
        
        _currentConnection.value = updatedConnection
    }
    
    /**
     * Получение сводки подключения
     */
    fun getConnectionSummary(): Map<String, Any> {
        val connection = _currentConnection.value
        val currentStatus = _status.value
        
        if (connection == null) {
            return mapOf(
                "isConnected" to false,
                "statusText" to currentStatus.displayName,
                "serverInfo" to emptyMap<String, Any>(),
                "connectionTime" to 0.0,
                "speed" to mapOf("download" to 0.0, "upload" to 0.0),
                "dataUsage" to mapOf("sent" to 0, "received" to 0, "total" to 0)
            )
        }
        
        val speed = connection.speed
        
        return mapOf(
            "isConnected" to (currentStatus == VPNStatus.CONNECTED),
            "statusText" to currentStatus.displayName,
            "serverInfo" to mapOf(
                "id" to connection.server.id,
                "name" to connection.server.name,
                "country" to connection.server.country,
                "flag" to connection.server.flag
            ),
            "connectionTime" to (connection.connectionTime / 1000.0),
            "speed" to mapOf(
                "download" to speed.first,
                "upload" to speed.second
            ),
            "dataUsage" to mapOf(
                "sent" to connection.bytesSent,
                "received" to connection.bytesReceived,
                "total" to (connection.bytesSent + connection.bytesReceived)
            )
        )
    }
    
    /**
     * Очистка ресурсов
     */
    fun cleanup() {
        scope.cancel()
        vpnInterface?.close()
        vpnInterface = null
        vpnService = null
    }
    
    // MARK: - Private Methods
    
    private fun loadServers() {
        scope.launch {
            try {
                // Загружаем серверы из API
                val servers = loadServersFromAPI()
                _availableServers.value = servers
            } catch (e: Exception) {
                // Используем серверы по умолчанию
                _availableServers.value = loadDefaultServers()
            }
        }
    }
    
    private suspend fun loadServersFromAPI(): List<VPNServer> {
        // Здесь будет загрузка серверов из API
        // Пока что возвращаем пустой список
        return emptyList()
    }
    
    private fun loadDefaultServers(): List<VPNServer> {
        return listOf(
            VPNServer(
                id = "sg-1",
                name = "Сингапур",
                country = "SG",
                flag = "🇸🇬",
                ip = "192.168.2.10",
                port = 443,
                protocol = VPNProtocol.SHADOWSOCKS,
                isAvailable = true,
                performanceScore = 95.0,
                ping = 25,
                load = 15
            ),
            VPNServer(
                id = "de-1",
                name = "Германия",
                country = "DE",
                flag = "🇩🇪",
                ip = "192.168.2.11",
                port = 443,
                protocol = VPNProtocol.V2RAY,
                isAvailable = true,
                performanceScore = 92.0,
                ping = 45,
                load = 25
            ),
            VPNServer(
                id = "hk-1",
                name = "Гонконг",
                country = "HK",
                flag = "🇭🇰",
                ip = "192.168.2.12",
                port = 443,
                protocol = VPNProtocol.SHADOWSOCKS,
                isAvailable = true,
                performanceScore = 88.0,
                ping = 35,
                load = 20
            ),
            VPNServer(
                id = "jp-1",
                name = "Япония",
                country = "JP",
                flag = "🇯🇵",
                ip = "192.168.2.13",
                port = 51820,
                protocol = VPNProtocol.WIREGUARD,
                isAvailable = true,
                performanceScore = 90.0,
                ping = 40,
                load = 30
            ),
            VPNServer(
                id = "us-1",
                name = "США",
                country = "US",
                flag = "🇺🇸",
                ip = "192.168.2.14",
                port = 1194,
                protocol = VPNProtocol.OPENVPN,
                isAvailable = true,
                performanceScore = 85.0,
                ping = 80,
                load = 40
            ),
            VPNServer(
                id = "ca-1",
                name = "Канада",
                country = "CA",
                flag = "🇨🇦",
                ip = "192.168.2.15",
                port = 51820,
                protocol = VPNProtocol.WIREGUARD,
                isAvailable = true,
                performanceScore = 87.0,
                ping = 75,
                load = 35
            )
        )
    }
    
    private suspend fun performConnection(server: VPNServer) {
        // Здесь будет реальная логика подключения к VPN
        // Пока что симулируем подключение
        
        delay(2000) // 2 секунды
        
        // В реальной реализации здесь будет настройка VpnService
        // и создание VPN туннеля
    }
    
    private suspend fun performDisconnection() {
        // Здесь будет реальная логика отключения от VPN
        // Пока что симулируем отключение
        
        delay(1000) // 1 секунда
        
        // В реальной реализации здесь будет закрытие VPN туннеля
        vpnInterface?.close()
        vpnInterface = null
    }
}

